src/SDKs/Batch/DataPlane/Azure.Batch/AddTasksWorkflowManager.cs

```diff
@@ -159,7 +160,8 @@ internal class AddTasksWorkflowManager
                 
                 //Attempt to take some items from the set of tasks remaining to be added and prepare it to be added
                 TrackedCloudTask taskToAdd;
-                while (nameToTaskMapping.Count < Constants.MaxTasksInSingleAddTaskCollectionRequest && this._remainingTasksToAdd.TryDequeue(out taskToAdd))
+                int tmpMaxTasks = this._maxTasks;
```

Comment:
Daryl had a nice comment about this is one of the Python SDK reviews. Basically it is possible for this to decrease from running threads failing, while this is looping.  
I beleive his argument was something along the lines of if this drops from 100->50 mid loop, expected behavior should be consistent with one of the max task sizes. There was also an argument for simplicity of code.
My initial argument was to not use this variable as current maxtasks is monotonically decreasing so worst case we get somewhere between the two. 
I am for reviewing removing it, but likely would need a review of the binary division to make sure everything sorts itself out.

Reply:
I believe one other argument Daryl had for python was uncertainty on atomic reads of ints in python 32 bit vs 64 bit.  That being said I believe int32 are atomic reads for either in C# or we could reduce to short as the max tasks per request in not likely to exceed that

Reply:
Long story short => variable makes it simpler. For large taskadds encourages 100->50->25->12->6->3->1 regression.
It will work either way.

Reply:
I don't think this really encourages 100->50->... as any time there is max degrees of parallelism involved, both legs are going to hit the error and both will do the division (right?).

With that said, this does dodge the issue of the value being updated when we're on the 77th task and then sending a request with a weird number of tasks. I'd like to see a comment explaining why we do this though as it reads weird at first glance, something like: `//Cache this in case a response comes in and reduces it while we're in the middle of building a new request`

Note that this would be another argument for the thing I mentioned in our office today of refactoring things to where errors are handled outside of the `StageFileAndAddTasks` method as I am pretty sure we could write things in such a way there was never any parallelism (i.e. no interlocked exchange needed and no wonky thing here either).

Not really a deliverable for you but figured I'd mention it.

Reply:
"I don't think this really encourages 100->50->... as any time there is max degrees of parallelism involved, both legs are going to hit the error and both will do the division (right?)." 
=> Currently behavior across the other sdk's (and current implementation in this PR) says when handling the AddTaskCollectionResponse(in which ever form) if requestEntityTooLarge, take requestsize.count/2 and iff this value is less than the current in memory size replace it. Ie: they are doing a division on themselves not on the variable used.

Reply:
"With that said, this does dodge the issue of the value being updated when we're on the 77th task and then sending a request with a weird number of tasks."
=> It dodges it if there are 100*degreesOfParallelism tasks being added. That being said behaves normally even, but per the success test case will cause possible 'weird' regression paths of 100->50->38->25(which i believe was the steady state).
Note if 38 was the steady state it would have been more optimal which was part of my initial argument not to include the cache